clippy_lints/src/missing_const_for_fn.rs

```diff
@@ -0,0 +1,113 @@
+use rustc::hir;
+use rustc::hir::{Body, FnDecl, Constness};
+use rustc::hir::intravisit::FnKind;
+// use rustc::mir::*;
+use syntax::ast::{NodeId, Attribute};
+use syntax_pos::Span;
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use rustc::{declare_tool_lint, lint_array};
+use rustc_mir::transform::qualify_min_const_fn::is_min_const_fn;
+use crate::utils::{span_lint, is_entrypoint_fn};
+
+/// **What it does:**
+///
+/// Suggests the use of `const` in functions and methods where possible
+///
+/// **Why is this bad?**
+/// Not using `const` is a missed optimization. Instead of having the function execute at runtime,
+/// when using `const`, it's evaluated at compiletime.
+///
+/// **Known problems:**
+///
+/// Const functions are currently still being worked on, with some features only being available
+/// on nightly. This lint does not consider all edge cases currently and the suggestions may be
+/// incorrect if you are using this lint on stable.
+///
+/// Also, the lint only runs one pass over the code. Consider these two non-const functions:
+///
+/// ```rust
+/// fn a() -> i32 { 0 }
+/// fn b() -> i32 { a() }
+/// ```
+///
+/// When running Clippy, the lint will only suggest to make `a` const, because `b` at this time
+/// can't be const as it calls a non-const function. Making `a` const and running Clippy again,
+/// will suggest to make `b` const, too.
+///
+/// **Example:**
+///
+/// ```rust
+/// fn new() -> Self {
+///     Self {
+///         random_number: 42
+///     }
+/// }
+/// ```
+///
+/// Could be a const fn:
+///
+/// ```rust
+/// const fn new() -> Self {
+///     Self {
+///         random_number: 42
+///     }
+/// }
+/// ```
+declare_clippy_lint! {
+    pub MISSING_CONST_FOR_FN,
+    nursery,
+    "Lint functions definitions that could be made `const fn`"
+}
+
+#[derive(Clone)]
+pub struct MissingConstForFn;
+
+impl LintPass for MissingConstForFn {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(MISSING_CONST_FOR_FN)
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for MissingConstForFn {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'_, '_>,
+        kind: FnKind<'_>,
+        _: &FnDecl,
+        _: &Body,
+        span: Span,
+        node_id: NodeId
+    ) {
+        let def_id = cx.tcx.hir().local_def_id(node_id);
+        let mir = cx.tcx.optimized_mir(def_id);
+        if let Err((span, err) = is_min_const_fn(cx.tcx, def_id, &mir) {
+            cx.tcx.sess.span_err(span, &err);
```

Comment:
I don't think we should run this before we know that the function can be a const fn.

Reply:
I'm pretty sure https://github.com/rust-lang/rust-clippy/pull/3648/commits/165d37d893fe6cd46c517cab14231303a9c565de is _not_ what you had in mind? If that's the case, could you elaborate a bit?

It currently triggers `projection without trait bound` ICEs in the dogfood tests in some external crates. I guess that's related, but I don't see how I could avoid that. If I remove the `span_err` call it will complain about not-emitted diagnostics.

Reply:
It is what I had in mind (mostly). The `tcx.is_min_const_fn` check is wrong, as that mostly just checks for stability and whether there's a `const` keyword infront of the `fn`

Reply:
> It currently triggers projection without trait bound ICEs in the dogfood tests in some external crates. I guess that's related, but I don't see how I could avoid that.

Do you have a repro or a backtrace? That seems like a bug in the `is_min_const_fn` function

Reply:
If anything the ICE is very weird: https://travis-ci.com/rust-lang/rust-clippy/jobs/170234558#L1247

Reply:
I think we should just remove the `delay_span_bug` in https://github.com/rust-lang/rust/blob/5012d7fb53d9f1e489c1ae1a081347fddbbffe5f/src/librustc_mir/transform/qualify_min_const_fn.rs#L36

We'll need to remove it anyway when we support trait bounds other than `Sized`, and I have not been able to find a way to trigger this ICE in regular Rust code, so we don't need any check there.

Reply:
`delay_span_bug` has been removed now :+1: